basin3d/core/monitor.py

```diff
@@ -0,0 +1,228 @@
+"""
+`basin3d.core.monitor`
+**********************
+
+.. currentmodule:: basin3d.core.monitor
+
+:synopsis: The BASIN-3D monitoring module
+:module author: Val Hendrix <vhendrix@lbl.gov>
+
+This module holds the logging functionality of BASIN-3D. It supports
+logging the messages and their contexts
+
+.. contents:: Contents
+    :local:
+    :backlinks: top
+
+"""
+import contextvars
+import logging
+import os
+import uuid
+from functools import wraps
+from logging import config
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import yaml
+
+LOGGER_NAME = __name__
+BASE_PATH = os.path.dirname(__file__)
+LOG_CONFIG_PATH = os.path.join(BASE_PATH, "logging.yaml")
+
+#: A unique identifier for a single request context. Use this
+#: If you want to identify log records from a single context
+#: This is a Context variable which is natively supported in asyncio
+#: and are ready to be used without any extra configuration.
+synthesis_id: contextvars.ContextVar = contextvars.ContextVar('synthesis_id')
+
+#: A unique identifier for a  basin3d context. Use this
+#: If you want to identify log records from a single context
+#: This is a Context variable which is natively supported in asyncio
+#: and are ready to be used without any extra configuration.
+basin3d_where: contextvars.ContextVar = contextvars.ContextVar('basin3d_where')
+
+
+class Basin3dLogger(logging.Logger):
+    """
+    Custom logger for adding realtime context information (basin3d_where)
+    """
+
+    def info(self, msg, *args, **kwargs):
+        kwargs = self._add_extra(**kwargs)
+        super().info(msg, *args, **kwargs)
+
+    def error(self, msg, *args, **kwargs):
+        kwargs = self._add_extra(**kwargs)
+        super().error(msg, *args, **kwargs)
+
+    def debug(self, msg, *args, **kwargs):
+        kwargs = self._add_extra(**kwargs)
+        super().debug(msg, *args, **kwargs)
+
+    def warning(self, msg, *args, **kwargs):
+        kwargs = self._add_extra(**kwargs)
+        super().warning(msg, *args, **kwargs)
+
+    def log(self, level, msg, *args, **kwargs):
+        kwargs = self._add_extra(**kwargs)
+        super().log(level, msg, *args, **kwargs)
+
+    def _add_extra(self, **kwargs):
+        """Add essdive extra, if exists """
```

Comment:
what is this variable for? and why is it for essdive (seems odd to put a data source in the core modules)?

Reply:
This allow you to add extra information to the log record which we need to do for BASIN-3D logging.  Note that it is a private method and not for publich consumption.

Reply:
OK. Then recommend taking out the reference to essdive.

Reply:
oh heh. That didn't even register.  LOL!  Now I know why you were asking